src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,274 @@
+//!    println!("listening on {}", addr);
+//!    axum_server::bind(addr)
+//!        .proxy_protocol_enabled()
+//!        .serve(app.into_make_service())
+//!        .await
+//!        .unwrap();
+//! }
+//! ```
+use std::future::poll_fn;
+use std::future::Future;
+use std::io;
+use std::net::IpAddr;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+use std::time::Duration;
+
+use http::HeaderValue;
+use http::Request;
+use ppp::{v2, HeaderResult, PartialResult};
+use tokio::io::AsyncReadExt;
+use tokio::io::ReadBuf;
+use tower_service::Service;
+
+use crate::accept::Accept;
+use std::fmt;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+pub mod future;
+use self::future::ProxyProtocolAcceptorFuture;
+
+pub trait Peekable {
+    fn poll_peek(
+        &mut self,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<std::io::Result<usize>>;
+}
+
+/// Note: Currently only supports the V2 PROXY header format.
+/// See proxy header spec: <https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt>
+pub(crate) async fn read_proxy_header<I>(mut stream: I) -> Result<(I, Option<IpAddr>), io::Error>
+where
+    I: Peekable + AsyncRead + Unpin,
+{
+    // Maximum expected size for PROXY header is
+    // unlikely to be larger than 512 bytes unless additional fields are added.
+    const BUFFER_SIZE: usize = 512;
+    // Mutable buffer for storing the bytes received from the stream.
+    let mut array_buffer = [0; BUFFER_SIZE];
+    let mut buffer = ReadBuf::new(&mut array_buffer);
+
+    // peek the stream until find a complete header
+    let header = loop {
+        // Peek the stream data into the buffer.
+        // Each loop re-reads the stream from beginning,
+        // but further in if more stream has arrived.
+        poll_fn(|cx| stream.poll_peek(cx, &mut buffer)).await?;
+
+        let header = HeaderResult::parse(buffer.filled());
+
+        if header.is_complete() {
+            break header;
+        } else if buffer.remaining() == 0 {
+            // Buffer limit reached without finding a complete header. Consider increasing the buffer size.
+            return Ok((stream, None));
+        }
+
+        buffer.clear();
+    };
+
+    match header {
+        HeaderResult::V1(Ok(_header)) => {
+            // V1 PROXY header detected, which is not supported
+            Ok((stream, None))
+        }
+        HeaderResult::V2(Ok(header)) => {
+            let client_address = match header.addresses {
+                v2::Addresses::IPv4(ip) => IpAddr::V4(ip.source_address),
+                v2::Addresses::IPv6(ip) => IpAddr::V6(ip.source_address),
+                v2::Addresses::Unix(_unix) => {
+                    // Unix socket addresses are not supported
+                    return Ok((stream, None));
+                }
+                v2::Addresses::Unspecified => {
+                    // V2 PROXY header addresses "Unspecified"
+                    return Ok((stream, None));
```

Comment:
Then why is this OK?

Reply:
this should be ok as it just means addresses are unspecified, and therefore in a forwarded header would be "unknown"